snuba/datasets/factory.py

```diff
@@ -1,93 +1,96 @@
-from typing import Callable, MutableMapping, Sequence, Set
+from typing import Generator, MutableMapping, Sequence, Type
 
-from snuba import settings
 from snuba.datasets.dataset import Dataset
 from snuba.util import with_span
+from snuba.utils.config_component_factory import ConfigComponentFactory
 from snuba.utils.serializable_exception import SerializableException
 
-DATASETS_IMPL: MutableMapping[str, Dataset] = {}
-DATASETS_NAME_LOOKUP: MutableMapping[Dataset, str] = {}
-
-DEV_DATASET_NAMES: Set[str] = set()
-
-DATASET_NAMES: Set[str] = {
-    "discover",
-    "events",
-    "groupassignee",
-    "groupedmessage",
-    "metrics",
-    "outcomes",
-    "outcomes_raw",
-    "sessions",
-    "transactions",
-    "profiles",
-    "functions",
-    "generic_metrics",
-    "replays",
-    *(DEV_DATASET_NAMES if settings.ENABLE_DEV_FEATURES else set()),
-}
+
+class _DatasetFactory(ConfigComponentFactory[Dataset, str]):
+    def __init__(self):
+        self._dataset_map: MutableMapping[str, Dataset] = {}
+        self._name_map: MutableMapping[Type[Dataset], str] = {}
+
+    def initialize(self):
+        from snuba.datasets.cdc.groupassignee import GroupAssigneeDataset
+        from snuba.datasets.cdc.groupedmessage import GroupedMessageDataset
+        from snuba.datasets.discover import DiscoverDataset
+        from snuba.datasets.events import EventsDataset
+        from snuba.datasets.functions import FunctionsDataset
+        from snuba.datasets.generic_metrics import GenericMetricsDataset
+        from snuba.datasets.metrics import MetricsDataset
+        from snuba.datasets.outcomes import OutcomesDataset
+        from snuba.datasets.outcomes_raw import OutcomesRawDataset
+        from snuba.datasets.profiles import ProfilesDataset
+        from snuba.datasets.replays import ReplaysDataset
+        from snuba.datasets.sessions import SessionsDataset
+        from snuba.datasets.transactions import TransactionsDataset
+
+        self._dataset_map.update(
+            {
+                "discover": DiscoverDataset(),
+                "events": EventsDataset(),
+                "groupassignee": GroupAssigneeDataset(),
+                "groupedmessage": GroupedMessageDataset(),
+                "metrics": MetricsDataset(),
+                "outcomes": OutcomesDataset(),
+                "outcomes_raw": OutcomesRawDataset(),
+                "sessions": SessionsDataset(),
+                "transactions": TransactionsDataset(),
+                "profiles": ProfilesDataset(),
+                "functions": FunctionsDataset(),
+                "generic_metrics": GenericMetricsDataset(),
+                "replays": ReplaysDataset(),
+            }
+        )
```

Comment:
I'm excited to the dataset factory in a cleaner state!

One question, is there a reason why the datasets mappings were changed to an object of Dataset as opposed to the class itself and returning a fresh instance in `get()`? Would this become problematic if we ever decide to call `select_entity` on one dataset, and later in the execution, use `get()` for a new query but same dataset? Not sure if that case exists.

Reply:
We discussed offline but TL;DR:

The old dataset code actually did use a map of strings to instances, just in a needlessly complicated way. That is now fixed